include/dr/shp/copy.hpp

```diff
@@ -75,19 +75,24 @@ template <std::forward_iterator InputIt, lib::distributed_iterator OutputIt>
   requires __detail::is_syclmemcopyable<std::iter_value_t<InputIt>,
                                         std::iter_value_t<OutputIt>>
 cl::sycl::event copy_async(InputIt first, InputIt last, OutputIt d_first) {
-  auto segments = lib::ranges::segments(d_first);
-  auto segment = std::begin(segments);
+  auto &&segments = lib::ranges::segments(d_first);
+  auto segment_iter = rng::begin(segments);
 
   std::vector<cl::sycl::event> events;
 
   while (first != last) {
-    const std::size_t n_to_copy =
-        std::min<size_t>((*segment).size(), last - first);
+    auto &&segment = *segment_iter;
+    auto size = std::distance(rng::begin(segment), rng::end(segment));
 
-    events.push_back(
-        shp::copy_async(first, first + n_to_copy, (*segment).begin()));
```

Comment:
Are we not allowed to use `iter + size`? If we can't - why? If we can - concise code is more readable to me.

Reply:
I would like the distributed copy functions to work for forward iterators, so I want to avoid using expressions like `iter + size` that only work on random access iterators.  `std::advance` will do the same thing, but work for all iterator types.